### PR TITLE
MapFileResolver returns null if asked for delisted tickers

### LINKED ISSUE #4049: MapResolver should return null if we ask for a delisted ticker
#### Expected Behavior

`MapResolver.ResolveMapFile(ticker, date)` should return `null ` if we ask for a ticker delisted before `date`

#### Actual Behavior

It returns the last `MapFile` with the ticker

#### Potential Solution

Check the `MapResolver.ResolveMapFile` date logic.

#### Reproducing the Problem

This test should pass, but it fails in current `master`.
```cs
[Test]
public void ResolvesDelistedSymbolAsNull()
{
	var mapResolver = new MapFileResolver(new List<MapFile>
	{
		new MapFile(
			"amz",
			new List<MapFileRow>
			{
				new MapFileRow(new DateTime(1998, 01, 02), "uwz"),
				new MapFileRow(new DateTime(1998, 09, 11), "uwz"),
				new MapFileRow(new DateTime(2004, 12, 31), "amz")
			}
		)
	});
	var date = new DateTime(2007, 01, 03);
	var ticker = "AMZ";
	Assert.IsNull(mapResolver.ResolveMapFile(ticker, date));
}
```

#### System Information

Windows

#### Checklist

- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [x] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by Jay-Jay-D**
#### Description

Update dates logic in `MapFileResolver.ResolveMapFile(symbol, date)` for cases where asked `date` is posterior to delisting date.

Also, check for `null `at `SecurityIdentifier.GetFirstTickerAndDate`. 

#### Related Issue

Closes #4049 

#### Motivation and Context

If we ask for a ticker for a date posterior to its delisting, `MapResolver` shouldn't return any map at all.

#### Requires Documentation Change

No.

#### How Has This Been Tested?

New test was added, all tests are green.

#### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Common/Data/Auxiliary/MapFileResolver.cs [hunk 1]
```diff
@@ -148,6 +148,7 @@ public MapFile ResolveMapFile(string symbol, DateTime date)
                     {
                         // the searched date is greater than the last date in the list, return the last entry
                         indexOf = entries.Keys.Count - 1;
+                        if (indexOf == 0) return null;
```
Comment (Martin-Molinero) [inline Common/Data/Auxiliary/MapFileResolver.cs]: Not sure this is right 🤔 - For the added unit test, add for example `new MapFileRow(new DateTime(2005, 12, 31), "amz")` and should still return `null`, but it doesn't, right?
I think it should directly return `null` in this `if` case, see comment on line 149

Comment (Jay-Jay-D) [reply Common/Data/Auxiliary/MapFileResolver.cs]: > add for example new MapFileRow(new DateTime(2005, 12, 31), "amz") and should still return

Rows in map files with duplicated tickers only makes sense in the first row (listing date) or in the last row (deslisting date or if it is still listed, 2050-12-31).

Adding a new row duplicating `AMZ` wouldn't be a correct test case.

Comment (Martin-Molinero) [reply Common/Data/Auxiliary/MapFileResolver.cs]: Agreed to directly return `null` on slack, pending change

**Comment by Jay-Jay-D** [comment]
The actual behavior, i.e. return the latest available map file even when delisted, was intended.

It was explicitly tested here:
https://github.com/QuantConnect/Lean/blob/835c99e55c2ff8ca41d694c608c56e336767d94c/Tests/Engine/DataFeeds/Auxiliary/MapFileResolverTests.cs#L134

Also, `algorithm.AddEquity("AAA", Resolution.Daily);` uses today as date for resolving the map file.

**Pull request closed without merging**